### ISSUE TITLE
/analyze may lead to cache store fail (*.nativecodeanalysis.xml not found)

### ISSUE THREAD

**Issue by Domain** [fastbuild/fastbuild#608]
I am not sure why some xxxx.nativecodeanalysis.xml file not generated.

**Comment by ffulin** [comment]
Can you please provide some more information to help identify the issue you are experiencing?:
 - version of FASTBuild being used
 - command line options being used to invoke FBuild.exe
 - Visual Studio compiler version being used
 - command line options (with -showcmds) for one of the failed objects
 - full text of error message
 - is the issue intermittent: does the same file works sometimes but not other times or does it consistently fail on certain files?
 - is the failure tied to the object being built locally or remotely? (are you seeing this problem while also using distributed compilation, without distributed compilation, or in both situations)?

**Comment by Domain** [comment]
I use the latest FASTBuild, i.e. 0.98, and visual studio 2017 15.9.8
And I know how to reproduce the problem:
Consider this simple c++ file:
`#include <string>
void f(const std::string& s) { (void)(s); }
void g() { f(0); }`
When I build this file locally, the compiler will give me a C6387 warning, and this will be treated as error since I turn on /WX switch, and there is a nativecodeanalysis.xml file generated.
Then I delete the obj and xml file. 
And I enable distributed compilation, no matter build locally or remotely (-forceremote), it will be passed, the obj is generated without nativecodeanalysis.xml file, which causes the cache store fail.
The compiler options is very simple: /nologo /c /WX /EHsc /analyze

**Comment by Domain** [comment]
Sorry, I am wrong.
The xml file will be generated when build locally in distributed mode. But no defects in the xml file. And cache store OK.
The xml file will not be generated when build remotely.

**Comment by ffulin** [comment]
Thankyou for the extra information - the repro case in particular was very useful as it turned out to be subtly important. I spent some time investigating the issue and was able to reproduce and fix the issues you saw, cause by two separate bugs. The fix (https://github.com/fastbuild/fastbuild/commit/466a542c3be2e42ae2063234d9787f356002b9f8) includes test coverage for these cases.

The fixes will be in the next release (v0.99). I'll leave this issue open until that version is released.

Some additional context follows:

**Static Analysis relying on SAL annotations broken when using -cache[read|write]**

Visual Studio static analysis detects some problems by just analyzing the code, while other problems require SAL annotations to hint to the analyzer (with the standard library being annotated for example). The example you gave (passing nullptr to std::string constructor) relies on these SAL annotations, while testing to date was relying on the other checks.

When using -cache[read|write] FASTBuild does two-pass compilation. When cl.exe is passed /E for the first pass in combination with /analyze, it seems to suppress the built in defines that activate SAL annotations, effectively disabling them. This is why building locally could result in an empty nativecodeanalysis.xml file in your example (no errors were detected).

The fix I have submitted ensures that the necessary define (_PREFAST_) is declared when preprocessing. This makes local static analysis consistent, whether -cache is used or not.

**Remote client not returning the nativecodeanalysis.xml**

An object file built remotely did not return the nativecodeanalysis.xml file to the local host. When we then tried to open this file to store to the cache, an error would occur.  The fix I have submitted adds the missing file to the payload returned to the local machine and unpacks it so the results of the analysis are seen and can be cached.

The distributed compilation protocol version has been bumped to ensure the next version will only communicate with workers with this fix.

**Cache polluted/version bumped**

Because of the first bug (where the SAL-based annotations are not working), objects compiled with -cachewrite mode would store the empty nativecodeanalysis.xml file. This means even with these fixes, incomplete analysis data could be retrieved from the cache. The cache version has been bumped in order to ensure existing bad cache data will be ignored.

**Comment by ffulin** [comment]
Further testing uncovered some more problems. The solution of adding _PREFAST_ doesn't work reliably (causes some very strange compile and link errors). It seems there is some undocumented compiler magic at work.

I've made a followup fix (https://github.com/fastbuild/fastbuild/commit/051bb22e39fdbcc5507828be5b0283100e7d9841) which disables distributed compilation for /analyze and documents this limitation. 

To summarize, in the next version (v0.99), objects with /analyze:
 - will be correctly/consistently stored/retrieved from the cache :)
 - will have all warnings/errors (SAL based or otherwise) present :)
 - will only be built locally :(

**Comment by Domain** [comment]
Oh, that's pity. Our project will cost more than 1 hour to build locally. I must disable analyze now.

**Comment by ffulin** [comment]
**> Oh, that's pity.**

Definitely. I hope that one day this can be made to work.

**Comment by ffulin** [comment]
FASTBuild v0.99 has now been released which the fixes mentioned above.
